src/Field/Definition/ChoiceField.php

```diff
@@ -57,30 +82,93 @@ protected function toManagementApiData () : array
 	 */
 	public function validateData (ComponentContext $context, array $contentPath, mixed $data, array $fullData) : void
 	{
-		$allowedValueTypeConstraints = new AtLeastOneOf([
-			new Type("string"),
-			new Type("int"),
-		]);
+		if ($this->allowMultiselect)
+		{
+			$this->validateMultiSelect($context, $contentPath, $data);
+		}
+		else
+		{
+			$this->validateSingleSelect($context, $contentPath, $data);
+		}
+	}
+
 
+	private function validateSingleSelect (ComponentContext $context, array $contentPath, mixed $data) : void
+	{
 		$context->ensureDataIsValid(
 			$contentPath,
 			$this,
 			$data,
 			[
-				$this->allowMultiselect
-					? new All([new NotNull(), $allowedValueTypeConstraints])
-					: $allowedValueTypeConstraints,
+				new AtLeastOneOf([
+					new Type("string"),
+					new Type("int"),
+				]),
+				$this->required
+					? new NotBlank()
```

Comment:
```suggestion
					? new NotNull()
```

Reply:
If you have an optional single-select choice and you didn't select anything, you'll get an empty string from the Storyblok API. `NotNull` won't be sufficient here, unless there is another case you're talking about.

Reply:
API design from hell

Reply:
Sadly, yes :(

I've just verified all use-cases I could think of (where Storyblok might be a bit weird) in various combinations of:

With the combinations of

- Optional/Required
- Empty (never filled out)
- Empty (filled out and then cleared
- Filled out
- Different choice types
	- self/static
	- stories
	- languages
	- external JSON
	- data source

And they all seem to work consistently coherent.

I can only assume that this API has been closely designed with a direct connection to the frontend where `<select>` will work with empty strings instead of `null`. Just a wild guess 🤷

Reply:
I've clarified some of the types and added an additional validation step so that we're enforcing required fields to not be empty (have at least one choice selected).